### PR TITLE
174 check class correctness

### LINKED ISSUE #174: Add typeid name/type registry to pack/unpack to confirm class correctness
@jstrzebonski Overall, the goal is to improve error checking in checkpoint. This issue is to register class IDs with the type register when we serialize to check if the same class is being traversed in the same order when de-serializing. Right now, we only register types for polymorphic classes because we need to reconstruct them based on runtime type information. In an error checking mode, we want to extend this registration to all classes and then put this information into the serialized buffer so we can check if the traversal matches the same types during de-serialization.

### PR DESCRIPTION
**Pull request by jstrzebonski**
Fixes #174 

TODO:
- [x] depending on error checking mode (ON/OFF) serialize types hashes and names recursively, or not
- [x] could the error print whole serialization path, not only the last wrong type?
- [x] move type_registry out of vrt namespace to a better place
- [x] set error checking flag in dockerfiles and ci/build_cpp.sh
- [x] clean up the code
- [x] add proper unit test

### CODE REVIEW

**Comment by jstrzebonski** [comment]
The error is not the prettiest, but I think is readable anyway:
"Unpacking wrong type, got=N10checkpoint5tests4unit11FailureObj1E, expected=N10checkpoint5tests4unit11FailureObj3E"

File: src/checkpoint/dispatch/dispatch.impl.h
Comment (jstrzebonski) [inline src/checkpoint/dispatch/dispatch.impl.h]: There's a lot of temporary code at the moment, but my general proposition (and IMHO the easiest one) is to simply throw from `withTypeIdx<T>()`, do a `try-catch` around `ap()` and for each catch add information about current type.

I added a temporary super simple test (`test_struct_ee`) to check this, and the result looks like this:

```
Unpacking wrong type, got=i, expected=i
path: N10checkpoint5tests4unit2AAE
path: N10checkpoint5tests4unit2BBE
path: N10checkpoint5tests4unit2CCE
path: N10checkpoint5tests4unit2DDE
path: N10checkpoint5tests4unit2EEE
```

What's left is to actually do this only if error checking is enabled. If it's not, `ap()` needs to be called without `try-catch` (and also whole `withTypeIdx()` should probably be erased).

Does it make any sense @lifflander @PhilMiller ?

Comment (lifflander) [reply src/checkpoint/dispatch/dispatch.impl.h]: Yes, this makes sense to me. Although I'm not sure I completely understand the necessity of `DoNothing`. Can you explain why you are checking for an `int` there?

Comment (jstrzebonski) [reply src/checkpoint/dispatch/dispatch.impl.h]: That's just temporary code to easily test that try-catch solution. I'll remove it soon.

**Comment by jstrzebonski** [comment]
With demangling added, the exception message is even better:
```
Unpacking wrong type, got=int, expected=int
path: checkpoint::tests::unit::AA
path: checkpoint::tests::unit::BB
path: checkpoint::tests::unit::CC
path: checkpoint::tests::unit::DD
path: checkpoint::tests::unit::EE
```

(And that `got=int, expected=int` is wrong obviously, but it's just for testing purpose).

File: src/checkpoint/dispatch/dispatch.impl.h
Comment (jstrzebonski) [inline src/checkpoint/dispatch/dispatch.impl.h]: That's the very basic check that only checks if `checkpoint::deserialize<T>(buffer)` is called for correct `T`. 
And there is no recursion for this check.

We wanted it to be enabled if "deep" checking is turned off.

File: .gitignore
Comment (PhilMiller) [inline .gitignore]: This sort of thing should probably be in one's own global git ignore configuration file, not in individual repositories.

Comment (jstrzebonski) [reply .gitignore]: OK, removed.

Comment (jstrzebonski) [reply .gitignore]: Although I guess that applies to `.dir-locals.el` as well. And if you take a look at vt's .gitignore there might be even more things to clean up :- )

File: src/checkpoint/dispatch/dispatch.impl.h
Comment (PhilMiller) [inline src/checkpoint/dispatch/dispatch.impl.h]: If the code is going to throw and catch for this, it should probably use a distinct exception type, so that we don't inadvertently catch some other unrelated exception that happens to arise e.g. in a user-written `serialize()` method.

Comment (PhilMiller) [reply src/checkpoint/dispatch/dispatch.impl.h]: With a custom exception type, the `path` lines should also include an incrementing number indicating the effective call depth.

**Comment by PhilMiller** [comment]
I like the logic. With some improvement to the error reporting, i think this will be ready to go.

File: src/checkpoint/dispatch/dispatch.impl.h
Comment (jstrzebonski) [reply src/checkpoint/dispatch/dispatch.impl.h]: I completely agree, thanks for spotting this!

Comment (jstrzebonski) [reply src/checkpoint/dispatch/dispatch.impl.h]: @PhilMiller done:

```
Unpacking wrong type, got=int (idx=3724543661), expected= (idx=1231123)
#0 int
#1 checkpoint::tests::unit::AA
#2 checkpoint::tests::unit::BB
#3 checkpoint::tests::unit::CC
#4 checkpoint::tests::unit::DD
#5 checkpoint::tests::unit::EE
```

File: docs/md/checkpoint_learn_build.md
Comment (jstrzebonski) [inline docs/md/checkpoint_learn_build.md:80]: It fixes this:
image

**Comment by jstrzebonski** [comment]
`docker-compose` constantly ends with error 137 on `nvcc-10.1, ubuntu, mpich` pipeline. That error indicates that the container run out of memory. @lifflander is there anything we can do about it?

Obviously, locally on my machine checkpoint compiles just fine in docker.

**Review by PhilMiller — APPROVED** [review]

**Comment by PhilMiller** [comment]
This all looks good to me, and the nvcc 10.1 failure seems to have been spurious, so I'm re-running it.

I'll ask that Jonathan and others get a chance at a final review pass before hitting 'Merge'

**Comment by PhilMiller** [comment]
Hrm, nope, I think compiling the Kokkos test with error checking is running nvcc out of memory

File: .github/workflows/dockerimage-nvcc-10-ubuntu-mpich.yml
Comment (PhilMiller) [inline .github/workflows/dockerimage-nvcc-10-ubuntu-mpich.yml]: Perhaps we should turn this one off.

Comment (jstrzebonski) [reply .github/workflows/dockerimage-nvcc-10-ubuntu-mpich.yml]: That's exactly what I wanted to try.

**Review by lifflander — APPROVED** [review]
This all looks good to me!

File: tests/unit/test_serialiation_error_checking.cc
Comment (PhilMiller) [inline tests/unit/test_serialiation_error_checking.cc]: I just noticed, there's a typo in the name of this file, and hence in the header

**Review by PhilMiller — CHANGES_REQUESTED** [review]

File: tests/unit/test_serialiation_error_checking.cc
Comment (jstrzebonski) [reply tests/unit/test_serialiation_error_checking.cc]: Renamed.

**Review by PhilMiller — APPROVED** [review]

**Pull request merged**
